Make custom sidebar fully moddable, and support user-defined sidebars

wapcaplet opened the pull request:
#### Summary

Features "Make custom sidebar fully moddable, and support user-defined sidebars"


#### Purpose of change

Build on #44683 to make the JSON-driven sidebar a first-class citizen in the sidebar options.

The way the "Custom" (json-driven) sidebar section works now, a "root layout" widget is defined for each of the two existing widths (narrow and wide), and this section is available in all four sidebar types (classic, compact, labels, and labels-narrow).

Now I believe the widget variable support is mature enough to consider adding a fifth (and subsequent) sidebar type, starting with "custom", entirely driven and rendered by `widgets.cpp` from "sidebar" JSON definitions. To make this work, widgets will need to integrate better with the `window_panel` class, which represents the labeled sections within the sidebar which we have previously allowed the user to toggle and rearrange.

Specifically, I want the "layout" style widgets (designed for grouping other widgets together) to be convertible to `window_panel` instances, with their builtin drawing callback function. This will allow "layout" widgets to render themselves in a "rows" arrangement (previously unsupported). I hope the integration with `window_panel` rendering will make it possible to draw other multi-line widgets such as the overmap and log output.


#### Describe the solution

- [x] Add a new "sidebar" style widget, with one "custom" sidebar defined and usable in-game
- [x] Add a `widget::get_window_panel` function.
- [x] Use window panels for drawing the root layout
- [x] Defer loading of custom window panel until after JSON is loaded (thanks @dseguin)
- [x] Get custom panel to render correctly for single-row layouts ("arrange": "columns")
- [x] Get custom panel to render correctly for multi-row layouts ("arrange": "rows")
- [x] Make total custom panel *width* customizable in JSON (perhaps via a `root_layout` attribute)
- [x] Provide hard-coded log, compass, minimap, and overmap (until they can be widget-ized)
- [x] Include help text on custom layout, with reference to sidebar JSON files
- [x] Refactor, cleanup, document
- [x] Update `doc/SIDEBAR_MOD.json` with notes on how this works, new "width" field for root layout, etc.

Link to updated [doc/SIDEBAR_MOD.md](https://github.com/CleverRaven/Cataclysm-DDA/blob/f97b8c948ad0a94603ef608d3556c41f373bcfc3/doc/SIDEBAR_MOD.md)


#### Describe alternatives you've considered

This PR went through a few iterations in collaboration with @dseguin - any alternatives that didn't work were rejected.

Beyond this PR, a few more important widgets are needed to make the new sidebar fully playable - these are addressed in #53874 


#### Testing

Use `}` to open the sidebar options, and select `custom` layout.

Use TAB to toggle widget layouts on and off; use `s` or Return to rearrange them.

Save and quit; reload and ensure custom panel settings are preserved.

Edit your `data/json/ui/sidebar.json`, to further customize your "custom" sidebar - the most straightforward customization being to change "width" - then save and reload your game to see the changes. Try adding other widgets to your main "sidebar" layout, or try editing the attributes of the existing widgets - for example, try changing the colors of the base "hitpoint_graph" widget used by all HP bars, or adding a widget that is not available in the default custom sidebar, like the encumbrance, warmth, or wetness widgets/layouts.


#### Additional context

Configuring sidebar with `}`, a new "custom" layout is available:

![image](https://user-images.githubusercontent.com/238395/147793908-29822325-0435-4835-a382-d24a3c8df47d.png)

The layout sections come from the `"style": "sidebar"` widgets defined `sidebar.json` (currently only one), along with a log and minimap, and a `Hint` at the top for how to customize:

![image](https://user-images.githubusercontent.com/238395/147793918-c49c588c-bfe9-4fb9-8139-c6904c0a7b53.png)

Because the "Log", "Compass", "Map", and "Overmap" sections have not yet been converted to widget representations, these four sections are automatically added to all custom "sidebar" based layouts. All of them provide essential functionality for gameplay that has no widget-based substitute, and I wanted to ensure their availability until there are such.

dseguin commented:
Hi, are you still thinking about working on this?

I was interested in getting this working, so I rebased the branch and updated it to use a callback. The callback defers loading the root_layout until the widgets are loaded. It just took a few changes: https://github.com/dseguin/Cataclysm-DDA/commit/4f83a6e05c8c0ff6b28cfa829317b950b7edb367

![custom_layout](https://user-images.githubusercontent.com/12537966/147145026-f7689bf4-181b-49e9-855e-d6e4eb432ab5.png)

wapcaplet commented:
> I was interested in getting this working, so I rebased the branch and updated it to use a callback. The callback defers loading the root_layout until the widgets are loaded. (...)
> Hopefully you haven't given up on this, because it's a cool feature!

Hey, thanks for lending a hand with this, I had rather given up on it.

I checked out your branch and got your patch merged with mine, but all it does is crash when I try to select "custom" from the sidebar configuration menu. Maybe I will have time to tinker with it over the holiday.

dseguin commented:
That's weird, I checked out your branch and it seems to be working:

https://user-images.githubusercontent.com/12537966/147184322-1a6a0dcc-1124-4f93-be1c-3b92c676cfd5.mp4

The only issue I'm finding is that the height of the widgets is always 0 in `game::draw_panels`, so the widget's draw function is never called. It looks like that might be the only issue left, though!

wapcaplet commented:
Are we on different platforms? It is nothing but segmentation faults for me.

- OS: Linux
    - OS Version: Distributor ID: Pop; Description: Pop!_OS 20.04 LTS; Release: 20.04; Codename: focal; 
- Game Version: 0.F-4544-g20291331a2 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    No Fungal Growth [no_fungal_growth],
    Bionic Professions [package_bionic_professions]
]

[crash.log](https://github.com/CleverRaven/Cataclysm-DDA/files/7766607/crash.log)

dseguin commented:
Pretty similar :P

```
- OS: Linux
    - OS Version: Distributor ID: Ubuntu; Description: Ubuntu 20.04.3 LTS; Release: 20.04; Codename: focal; 
- Game Version: 0.F-4544-g20291331a2 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    No Fungal Growth [no_fungal_growth],
    Bionic Professions [package_bionic_professions]
]

```

Here's the full make command that I use to compile:
```
make -j$(nproc) TILES=1 SOUND=1 LOCALIZE=1 RELEASE=0 CCACHE=1 SANITIZE=address
```

wapcaplet commented:
By passing a height of 1 in the `get_window_panel` call in `initialize_default_custom_panels`, I can get something to show up. I don't remember why I made the default value 0; maybe 1 would be better. Still doesn't quite work:

![image](https://user-images.githubusercontent.com/238395/147285878-b411c220-777d-4527-ba69-b3506002f72f.png)

Toggling the sections on and off seems to work, but when I return to the game, it crashes. If I'm lucky, there will be an error or two along the way:

```
 DEBUG    : Could not find bodypart bp_null in Reginald Brothers's body
 FUNCTION : const bodypart *Creature::get_part(const bodypart_id &) const
 FILE     : src/creature.cpp
 LINE     : 1942

 DEBUG    : Couldn't find color 2036295013
 FUNCTION : color_id color_manager::color_to_id(const nc_color &) const
 FILE     : src/color.cpp
 LINE     : 134
```
[crash.log](https://github.com/CleverRaven/Cataclysm-DDA/files/7771145/crash.log)

So, a few more hurdles to getting this working. I will keep hacking at it, and will welcome any more help you can provide. Feel free to submit a pull request directly to my branch.

dseguin commented:
Exciting news: the crashing/instability is fixed! For some reason, I can't make pull requests directly to your fork, so just check out the last 2 commits from https://github.com/dseguin/Cataclysm-DDA/tree/w-widget-window-panel

It turns out the issue was that `draw_func` lambda. The parent widget is null by the time the lambda is called, so it crashes as soon as `_style` is dereferenced. I fixed it by moving the lambda out into a static function. It's janky but it works. The only issue now is making the height dynamic:

![working_custom_sidebar](https://user-images.githubusercontent.com/12537966/147302065-9de3cb2f-95d3-45a7-9d5f-29aaa6482212.png)

wapcaplet commented:
Thanks again for providing a working patch. I would prefer a less invasive approach than adding `widget *` to the signature of all the panel methods, but it's more important to have something that works for now - it can be tidied up later.

dseguin commented:
No problem. Sorry about the mess, haha. I'll work on fixing that.

wapcaplet commented:
> No problem. Sorry about the mess, haha. I'll work on fixing that.

@dseguin You may want to hold off for now, I am doing a clean rebase against master now that the mood-constification was merged. This version is still working really well so far.

wapcaplet commented:
Okay, and while I'm having a great time revamping the JSONs for the sidebar layouts, those commits are largely off-topic for this PR. I will separate those to keep this PR more focused.

dseguin commented:
If you get the chance, take a look at https://github.com/wapcaplet/Cataclysm-DDA/pull/2. I managed to get the draw call parameters cleaned up into a nice central object (along with some minor `const` fixes as well). Hopefully that addresses the messiness!

wapcaplet commented:
> If you get the chance, take a look at [wapcaplet#2](https://github.com/wapcaplet/Cataclysm-DDA/pull/2). I managed to get the draw call parameters cleaned up into a nice central object (along with some minor `const` fixes as well). Hopefully that addresses the messiness!

Looking closer now, I see what you did there. Nice solution! I changed `_a` to `_ava` and `_w` to `_win` to make them more self-documenting, and fit with `_wgt`.

Now I am curious if this would have any effect on the lambda problem we had before...

wapcaplet commented:
Something else I would *love* to do, maybe not in this PR but soon, is a new widget style "sidebar", replacing the hard-coded "root layout" widgets with something more versatile, e.x.

```json
{
  "id": "custom_sidebar_a",
  "type": "widget",
  "style": "sidebar",
  "width": 50,
  "label": "custom A",
  "widgets": [ "hitpoints_layout", "stats_layout", ... ]
},
{
  "id": "custom_sidebar_b",
  "type": "widget",
  "style": "sidebar",
  "width": 80,
  "label": "custom B",
  "widgets": [ "hitpoints_xx_wide_layout", "stats_xx_wide_layout", ... ]
}
```

These "custom A" and "custom B" sidebars would be listed at the same level as "classic", "compact", "labels", and "labels narrow", and their widgets would be toggleable/sortable like we've done here with window panel callbacks.

Where I see this going is the ability to create completely custom sidebars in a modular (and moddable) fashion, which can still leverage the existing toggle/sort customization feature.

I've started cracking open this particular egg, and it looks like it may be feasible to do here in this PR. We've come this far, it may be worth going all the way!

dseguin commented:
> Something else I would love to do, maybe not in this PR but soon, is a new widget style "sidebar", replacing the hard-coded "root layout" widgets with something more versatile [...]

That's awesome, I'm all for it! :+1:

Let me know if you think of something I can help with.

wapcaplet commented:
It works! And it was not that hard - mostly finding the right place to call `panel_manager::init()` so the widgets will have been loaded (just after `load_world_modfiles()`, turns out), then adding a `widget::get_all()` and picking out the "sidebar" ones.

Bonus, this means we don't need the somewhat awkward `_load_fn` workaround you added to `panel_layout`, since the panel initialize-default functions have ready access to the widget instances - so I took that out.

I think it is looking pretty solid, but would love if you can do some playtesting and try messing around with your sidebar JSONs. The visual impact of this PR is rather meh right now, so it would be nice to flesh out this new custom sidebar with more sections, and/or provide a second sidebar at a different width, so if you come up with some cool arrangements, let me know.

dseguin commented:
I tried this out with different widths and widgets. No issues at all. The code looks so much better too!

In fact, I went and reduced the width limit to 16 on my side just to see how it looks, and surprisingly it's not bad:
![super_narrow](https://user-images.githubusercontent.com/12537966/147525434-67f1c4ac-f6a3-413e-b38a-f805f998476a.png)

This is a real game changer. Thanks for seeing this through!

Inglonias commented:
Wow. Ok. If this gets enough support, I bet we could replace the sidebar infrastructure entirely with widgets (Eventually)

wapcaplet commented:
> Wow. Ok. If this gets enough support, I bet we could replace the sidebar infrastructure entirely with widgets (Eventually)

That has been my long-term plan, and we are *nearly* there. I would love to someday remove most of the hard-coded stuff in `panels.cpp`. A few other "must-haves" are support for the Map, Log, and Compass (widgets can't do these yet), and some indicators for bleeding/infected body parts (the HP widgets don't do this yet), along with any other bits of info that still need widget-izing - vehicle speed/direction/fuel, sky/weather, and movement mode indicator are the major ones I can think of.

I've been tinkering with this for more than a year, so it's exciting to see it finally coming together.

wapcaplet commented:
I did a few passes of rebasing to clean up and squash commits, and want to go back and tidy up the commit messages a bit more, but otherwise I think this is close to being ready.

wapcaplet commented:
I hemmed and hawed over the PR title, since this is more of a game-changer than I originally planned, but I think "Make custom sidebar fully moddable" definitely sounds better than "Use window panel for layout widgets".

Inglonias commented:
wow.

TadhgEomonn commented:
Mana from Magiclysm has been neglected.
